File: tools/clod-poc/src/qa/unified/region_probes.ts
import { sobelMagnitudes } from "./edge_metrics.js";
import { rec709Luminance, type LinearImage } from "./image_linear.js";
import type { QaRegionProbe } from "./schema.js";

export interface RegionProbeMetrics {
  luminanceMean: number;
  luminanceStddev: number;
  luminanceMin: number;
  luminanceMax: number;
  luminanceP05: number;
  luminanceP50: number;
  luminanceP95: number;
  chromaMean: number;
  blackPixelFraction: number;
  clippedPixelFraction: number;
  meanRgb: [number, number, number];
  edgeMagnitude: number;
}

export interface RegionProbeResult {
  id: string;
  status: "PASS" | "FAIL";
  metrics: RegionProbeMetrics;
  failures: string[];
}

export function evaluateRegionProbe(
  image: LinearImage,
  probe: QaRegionProbe,
): RegionProbeResult {
  const [rx, ry, rw, rh] = probe.rect_normalized;
  const x0 = Math.floor(rx * image.width);
  const y0 = Math.floor(ry * image.height);
  const x1 = Math.max(x0 + 1, Math.ceil((rx + rw) * image.width));
  const y1 = Math.max(y0 + 1, Math.ceil((ry + rh) * image.height));
  const edges = sobelMagnitudes(image);
  const luminance: number[] = [];
  let sumR = 0;
  let sumG = 0;
  let sumB = 0;
  let chroma = 0;
  let black = 0;
  let clipped = 0;
  let edge = 0;
  for (let y = y0; y < Math.min(y1, image.height); y++) {
    for (let x = x0; x < Math.min(x1, image.width); x++) {
      const pixel = y * image.width + x;
      const offset = pixel * 3;
      const r = image.rgb[offset] ?? 0;
      const g = image.rgb[offset + 1] ?? 0;
      const b = image.rgb[offset + 2] ?? 0;
      const yValue = rec709Luminance(r, g, b);
      luminance.push(yValue);
      sumR += r;
      sumG += g;
      sumB += b;
      chroma += Math.max(r, g, b) - Math.min(r, g, b);
      if (yValue < 0.01) black++;
      if (r > 0.99 || g > 0.99 || b > 0.99) clipped++;
      edge += edges[pixel] ?? 0;
    }
  }
  if (luminance.length === 0) {
    throw new Error(`region probe ${probe.id} selected no pixels`);
  }
  luminance.sort((a, b) => a - b);
  const count = luminance.length;
  const luminanceMean = luminance.reduce((sum, value) => sum + value, 0) / count;
  const metrics: RegionProbeMetrics = {
    luminanceMean,
    luminanceStddev: Math.sqrt(
      luminance.reduce((sum, value) => sum + (value - luminanceMean) ** 2, 0) / count,
    ),
    luminanceMin: luminance[0] ?? 0,
    luminanceMax: luminance[count - 1] ?? 0,
    luminanceP05: percentile(luminance, 0.05),
    luminanceP50: percentile(luminance, 0.50),
    luminanceP95: percentile(luminance, 0.95),
    chromaMean: chroma / count,
    blackPixelFraction: black / count,
    clippedPixelFraction: clipped / count,
    meanRgb: [sumR / count, sumG / count, sumB / count],
    edgeMagnitude: edge / count,
  };
  const failures: string[] = [];
  gate("luminance_mean", metrics.luminanceMean, probe.gates.luminance_mean, failures);
  gate(
    "luminance_stddev",
    metrics.luminanceStddev,
    probe.gates.luminance_stddev,
    failures,
  );
  gate("chroma_mean", metrics.chromaMean, probe.gates.chroma_mean, failures);
  gate(
    "black_pixel_fraction",
    metrics.blackPixelFraction,
    probe.gates.black_pixel_fraction,
    failures,
  );
  gate(
    "clipped_pixel_fraction",
    metrics.clippedPixelFraction,
    probe.gates.clipped_pixel_fraction,
    failures,
  );
  gate("edge_magnitude", metrics.edgeMagnitude, probe.gates.edge_magnitude, failures);
  return {
    id: probe.id,
    status: failures.length === 0 ? "PASS" : "FAIL",
    metrics,
    failures,
  };
}

function percentile(sorted: readonly number[], quantile: number): number {
  return sorted[Math.min(sorted.length - 1, Math.floor((sorted.length - 1) * quantile))] ?? 0;
}

function gate(
  name: string,
  value: number,
  range: { min?: number; max?: number } | undefined,
  failures: string[],
): void {
  if (!range) return;
  if (range.min !== undefined && value < range.min) {
    failures.push(`${name} ${value} < ${range.min}`);
  }
  if (range.max !== undefined && value > range.max) {
    failures.push(`${name} ${value} > ${range.max}`);
  }
}

File: src/diagnostics/qa/unified/region_probes.rs
use serde::Serialize;

use super::edge_metrics::sobel_magnitudes;
use super::image_linear::{LinearImage, rec709_luminance};
use super::schema::{NumericRange, RegionProbe};

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct RegionProbeMetrics {
    pub luminance_mean: f64,
    pub luminance_stddev: f64,
    pub luminance_min: f64,
    pub luminance_max: f64,
    pub luminance_p05: f64,
    pub luminance_p50: f64,
    pub luminance_p95: f64,
    pub chroma_mean: f64,
    pub black_pixel_fraction: f64,
    pub clipped_pixel_fraction: f64,
    pub mean_rgb: [f64; 3],
    pub edge_magnitude: f64,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct RegionProbeResult {
    pub id: String,
    pub status: RegionStatus,
    pub metrics: RegionProbeMetrics,
    pub failures: Vec<String>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum RegionStatus {
    Pass,
    Fail,
}

impl RegionStatus {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Pass => "PASS",
            Self::Fail => "FAIL",
        }
    }
}

pub fn evaluate_region_probe(image: &LinearImage, probe: &RegionProbe) -> RegionProbeResult {
    let [normalized_x, normalized_y, normalized_width, normalized_height] = probe.rect_normalized;
    let x0 = (normalized_x * f64::from(image.width)).floor() as u32;
    let y0 = (normalized_y * f64::from(image.height)).floor() as u32;
    let x1 = ((normalized_x + normalized_width) * f64::from(image.width))
        .ceil()
        .max(f64::from(x0 + 1)) as u32;
    let y1 = ((normalized_y + normalized_height) * f64::from(image.height))
        .ceil()
        .max(f64::from(y0 + 1)) as u32;
    let x1 = x1.min(image.width);
    let y1 = y1.min(image.height);
    let edges = sobel_magnitudes(image);

    let mut luminance = Vec::new();
    let mut sum_rgb = [0.0_f64; 3];
    let mut chroma_sum = 0.0;
    let mut black_count = 0_u64;
    let mut clipped_count = 0_u64;
    let mut edge_sum = 0.0;

    for y in y0..y1 {
        for x in x0..x1 {
            let pixel = y as usize * image.width as usize + x as usize;
            let offset = pixel * 3;
            let rgb = [
                image.rgb[offset],
                image.rgb[offset + 1],
                image.rgb[offset + 2],
            ];
            let value = rec709_luminance(rgb[0], rgb[1], rgb[2]);
            luminance.push(f64::from(value));
            for channel in 0..3 {
                sum_rgb[channel] += f64::from(rgb[channel]);
            }
            chroma_sum += f64::from(
                rgb.iter().copied().fold(f32::NEG_INFINITY, f32::max)
                    - rgb.iter().copied().fold(f32::INFINITY, f32::min),
            );
            if value < 0.01 {
                black_count += 1;
            }
            if rgb.iter().any(|channel| *channel > 0.99) {
                clipped_count += 1;
            }
            edge_sum += f64::from(edges[pixel]);
        }
    }

    luminance.sort_by(|left, right| left.total_cmp(right));
    let count = luminance.len().max(1);
    let luminance_mean = luminance.iter().sum::<f64>() / count as f64;
    let luminance_stddev = (luminance
        .iter()
        .map(|value| {
            let delta = value - luminance_mean;
            delta * delta
        })
        .sum::<f64>()
        / count as f64)
        .sqrt();
    let metrics = RegionProbeMetrics {
        luminance_mean,
        luminance_stddev,
        luminance_min: luminance.first().copied().unwrap_or(0.0),
        luminance_max: luminance.last().copied().unwrap_or(0.0),
        luminance_p05: percentile(&luminance, 0.05),
        luminance_p50: percentile(&luminance, 0.50),
        luminance_p95: percentile(&luminance, 0.95),
        chroma_mean: chroma_sum / count as f64,
        black_pixel_fraction: black_count as f64 / count as f64,
        clipped_pixel_fraction: clipped_count as f64 / count as f64,
        mean_rgb: [
            sum_rgb[0] / count as f64,
            sum_rgb[1] / count as f64,
            sum_rgb[2] / count as f64,
        ],
        edge_magnitude: edge_sum / count as f64,
    };

    let mut failures = Vec::new();
    for (metric, range) in &probe.gates {
        let observed = match metric.as_str() {
            "luminance_mean" => metrics.luminance_mean,
            "luminance_stddev" => metrics.luminance_stddev,
            "chroma_mean" => metrics.chroma_mean,
            "black_pixel_fraction" => metrics.black_pixel_fraction,
            "clipped_pixel_fraction" => metrics.clipped_pixel_fraction,
            "edge_magnitude" => metrics.edge_magnitude,
            _ => continue,
        };
        evaluate_range(metric, observed, range, &mut failures);
    }

    RegionProbeResult {
        id: probe.id.clone(),
        status: if failures.is_empty() {
            RegionStatus::Pass
        } else {
            RegionStatus::Fail
        },
        metrics,
        failures,
    }
}

fn percentile(sorted: &[f64], quantile: f64) -> f64 {
    if sorted.is_empty() {
        return 0.0;
    }
    let index = ((sorted.len() - 1) as f64 * quantile).floor() as usize;
    sorted[index.min(sorted.len() - 1)]
}

fn evaluate_range(
    metric: &str,
    observed: f64,
    range: &NumericRange,
    failures: &mut Vec<String>,
) {
    if let Some(minimum) = range.min
        && observed < minimum
    {
        failures.push(format!("{metric} {observed} < {minimum}"));
    }
    if let Some(maximum) = range.max
        && observed > maximum
    {
        failures.push(format!("{metric} {observed} > {maximum}"));
    }
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeMap;

    use super::*;

    #[test]
    fn evaluates_signal_and_variance_boundaries() {
        let image = LinearImage {
            width: 2,
            height: 1,
            rgb: vec![0.0, 0.0, 0.0, 1.0, 1.0, 1.0],
        };
        let probe = RegionProbe {
            id: "signal".into(),
            rect_normalized: [0.0, 0.0, 1.0, 1.0],
            gates: BTreeMap::from([
                (
                    "luminance_mean".into(),
                    NumericRange {
                        min: Some(0.49),
                        max: Some(0.51),
                    },
                ),
                (
                    "luminance_stddev".into(),
                    NumericRange {
                        min: Some(0.49),
                        max: None,
                    },
                ),
            ]),
        };
        let result = evaluate_region_probe(&image, &probe);
        assert_eq!(result.status, RegionStatus::Pass);
    }
}
